### PR TITLE
refactor: remove composite action runcontext workaround

### PR DESCRIPTION
**Pull request by KnisterPeter**
The RunContext is cloned to execute a composite action with all its
steps in a similar context. This required some workaround, since
the command handler has kept a reference to the original RunContext.

This is solved now, by replacing the docker LogWriter with a proper
scoped LogWriter.

This prepares for a simpler setup of composite actions to be able
to create and re-create the composite RunContext for pre/main/post
action steps.

### CODE REVIEW

**Comment by ChristopherHX** [comment]
We need more tests for nodejs and docker actions before this PR can be merged, or we see heavy regressions.
I think the following is broken now:
- GITHUB_ACTION_PATH for nodejs and docker actions
- ${{ github.action_path }} for nodejs and docker actions

I only have added tests for GITHUB_ACTION_PATH in composite actions, which seem to be still working if I can trust my tests.

**Comment by KnisterPeter** [comment]
I will add tests tomorrow

**Comment by KnisterPeter** [comment]
According to the [docs](https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables) the `GITHUB_ACTION_PATH` is only supported in composite actions.

I'll add tests for `GITHUB_ACTION` and `GITHUB_ACTION_REPOSITORY`

**Comment by ChristopherHX** [comment]
Ok, `GITHUB_ACTION_PATH` is only for composite.
What about `${{github.action_ref}}` for docker actions?
`GITHUB_ACTION_REF` seems not to be documented, but provided.

**Comment by KnisterPeter** [comment]
I will add tests for `GITHUB_ACTION_REF` as well

**Comment by KnisterPeter** [comment]
@nektos/act-maintainers I'd consider this refactoring fine now. It already fixes regressions related to the env-var in actions.
This changes makes creating the `RunContext` for composite actions simpler. I will further refactor if this is merged, to have the `pre` and `post` steps available before and after the action is executed.

**Review by ChristopherHX — COMMENTED** [review]
Hmm I have no idea why I don't see any log output of run steps inside composite actions.
Due to this I couldn't test if `::add-mask::` is still working.

File: pkg/runner/action.go
Comment (KnisterPeter) [reply pkg/runner/action.go:397]: Indeed that is strange. I didn't change the logger, just the command handler wrapped around the log output. I will investigate this.

Comment (KnisterPeter) [reply pkg/runner/action.go]: Should be fixed

**Review by ZauberNerd — APPROVED** [review]

**Comment by ChristopherHX** [comment]
I have issues with `::add-mask::` inside composite actions. If you echo the secret value after the command your secret get exposed to the log, while in v0.2.26 it is replaced by `***`.

This make me think we need tests which tests secret masking.

**Comment by KnisterPeter** [comment]
Right. I will add a test for that tomorrow

**Comment by KnisterPeter** [comment]
I had to rebase to resolve conflicts with master.

**Comment by catthehacker** [comment]
> well catthehacker has done weird stuff with installing node that it is not in PATH while using actions/runner.

?

**Comment by ChristopherHX** [comment]
>> well catthehacker has done weird stuff with installing node that it is not in PATH while using actions/runner.
>
>?

I try to explain my statement

> docker inspect f486756abf2f --format "{{range .Config.Env}}{{println .}}{{end}}"
RUNNER_TOOL_CACHE=/opt/hostedtoolcache
RUNNER_OS=Linux
RUNNER_TEMP=/tmp
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
DEBIAN_FRONTEND=noninteractive

PATH has no node, this command is used by actions/runner to get the container env.

> docker run --rm catthehacker/ubuntu:act-latest node
docker: Error response from daemon: OCI runtime create failed: container_linux.go:367: starting container process caused: exec: "node": executable file not found in $PATH: unknown.

docker cannot find it, only act can. I guess act reads /etc/environment, but docker and actions/runner doesn't.

**Comment by catthehacker** [comment]
> PATH has no node, this command is used by actions/runner to get the container env.

Because it's not possible to do that automatically, I would have to manually write it into Dockerfile.

> docker cannot find it, only act can. I guess act reads /etc/environment, but docker and actions/runner doesn't.

I can add symlink to `/usr/bin` for `node`, that would fix it

**Comment by KnisterPeter** [comment]
> .....................................................
> Your composite workaround removal PR is causing more and more regressions for me :(
> - Now the composite has it's own job logger forced to stdout
>   - the color of the composite steps doesn't match it's job
>   - no longer a custom logger for me :(

You are referring to regressions in your fork? Or where do you see that?
The color has already a workaround. I think that may need some additional, more clean, work.

The logging is kind of a messy setup, hard to test, hard to change without regressions.
We should be able to reconfigure it with the command handler without recreating the logger. But that would need even more refactoring.

Should we do it upfront? In another PR?

**Comment by KnisterPeter** [comment]
> I can add symlink to `/usr/bin` for `node`, that would fix it

I think that is a good idea

**Comment by ChristopherHX** [comment]
> You are referring to regressions in your fork?

Oh it was https://github.com/nektos/act/pull/1110 in, I thought it had all changes from here....

> node actions cannot find node in PATH in act-latest image inside composite actions

If you fix this one I, **without symlink to /usr/bin for node**. I approve this change.
> In another PR?

That's ok, if it is merged within a month after this one, or I have to freeze github-act-runner for a long time.

**Comment by ChristopherHX** [comment]
> You are referring to regressions in your fork?

Whoops this commit has the color bug on my machine
https://github.com/nektos/act/pull/1089/commits/b082c4f289b6ce3beadce180c7f56ff26601a251

**Comment by KnisterPeter** [comment]
> > You are referring to regressions in your fork?
> 
> Whoops this commit has the color bug on my machine [b082c4f](https://github.com/nektos/act/commit/b082c4f289b6ce3beadce180c7f56ff26601a251)

Maybe the `keepColor` param I added here (2d06b71a2c9d5a524ce19a08b50ce500aa5a7647) isn't working as expected.
I guess it is switching colors when multiple jobs are started in parallel and then in one of them (not the latest) a composite action is executed. It will keep the color of the 'last' job.
I really think we should overhaul the logger, but I will try to workaround this on this PR before going into the logger full-time.

**Comment by KnisterPeter** [comment]
I pushed a new commits (856671e4901790831c6349ed6349e039ad32e1df) with a new workaround the mentioned issues. The tests should still be fine.

- masking does work
- coloring does work
- no new logger is created

@ChristopherHX Please check again if you don't mind.

File: pkg/runner/action.go
Comment (ChristopherHX) [inline pkg/runner/action.go:416]: You forget the extrapath? My clone method would create a flat nearly 1:1 copy even after someone would add new properties.
```suggestion
			Masks:            rc.Masks,
			ExtraPath:        rc.ExtraPath,
```

File: pkg/runner/action.go
Comment (ChristopherHX) [inline pkg/runner/action.go:449]: I forget the extrapath here too, while creating my initial PR. No reviewer complained :(.
```suggestion
		rc.Masks = compositerc.Masks
		rc.ExtraPath = compositerc.ExtraPath
```

File: pkg/runner/logger.go
Comment (ChristopherHX) [inline pkg/runner/logger.go:78]: > coloring does work

You disabled coloring all jobs are blue for me, by never doing nextcolor++. You are correct that composite actions have the same color in this PR.
```suggestion
	nextColor++
	ctx = WithMasks(ctx, masks)
```

**Review by ChristopherHX — COMMENTED** [review]
Nice, but still some bugs to address. Seems like we should write tests and no new code.
If you apply my suggestions, then I think this is ready to be merged.

File: pkg/runner/action.go
Comment (KnisterPeter) [reply pkg/runner/action.go:389]: 👍

File: pkg/runner/action.go
Comment (KnisterPeter) [reply pkg/runner/action.go:416]: Yes, I missed that. I removed the clone method, because I felt a bit unsure on that. It did a shallow clone and keept most attributes of the struct. That means it was possible to manipulate references from the clone to the outer RC.
This could have been done very easy without notice. Having this more explicit might lead to miss an update here, but that is easier to track down.

Comment (KnisterPeter) [reply pkg/runner/action.go:416]: If you think it's better to clone, I could re-add it.

File: pkg/runner/action.go
Comment (KnisterPeter) [reply pkg/runner/action.go:389]: > This fixes the node not found in composite issue

Can you point me to the issue, I would like to write a test for that.

Comment (KnisterPeter) [reply pkg/runner/action.go:389]: Awesome, I will look into that :+1:

**Review by ChristopherHX — DISMISSED** [review]
Thanks, everything from my side against merging this has been addressed. Works fine on my device on windows.

**Comment by KnisterPeter** [comment]
@cplee This is waiting for a review/merge.

**Review by cplee — APPROVED** [review]
works for me locally ✅

**Pull request merged**
